### PR TITLE
Introduce enums for energy and virial flags

### PR DESCRIPTION
**Pull request by rbberger**
**Summary**

Define energy and virial constants in `force.h` to reduce magic numbers in code.
Also defines constants `CENTROID_SAME`, `CENTROID_AVAIL` and `CENTROID_NOTAVAIL`.

**Related Issue(s)**

#2469 

**Author(s)**

Richard Berger (Temple U) @rbberger 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

File: src/pair.cpp
Comment (yafshar) [inline src/pair.cpp:790]: @rbberger This is creating the wrong value. Similar to what happened at #2469

```cpp
eflag=0 -> 0
eflag=1 -> 0
eflag=2 -> 2
eflag=3 -> 2
```
It should be something like `bool(i & ENERGY_PER_ATOM)`

File: src/pair.cpp
Comment (yafshar) [inline src/pair.cpp:793]: @rbberger the same thing here

File: src/pair.cpp
Comment (akohlmey) [reply src/pair.cpp:790]: > @rbberger This is creating the wrong value. Similar to what happened at #2469

the value doesn't matter. all the tests are for zero vs. non-zero.

[...]

> It should be something like `bool(i & ENERGY_PER_ATOM)`

if you absolutely want 0 vs 1 it should be more like `(eflag & ENERGY_PER_ATOM) ? 1 : 0`
those flags are "int" not "bool".

File: src/pair.h
Comment (akohlmey) [inline src/pair.h:22]: some of those constants are rather long names. not sure that helps with readability.
how about this?
- ENERGY_NONE
- ENERGY_GLOBAL
- ENERGY_PERATOM
- VIRIAL_TALLY
- VIRIAL_FDOTR
- VIRIAL_PERATOM
- VIRIAL_CENTROID

File: src/pair.cpp
Comment (yafshar) [reply src/pair.cpp:790]: @akohlmey I fixed Kim to avoid the requirement for value. Just mentioned the value for the backward compatibility and if there is any other piece of code that checks the value otherwise you are right, the value is not important

File: src/pair.h
Comment (rbberger) [reply src/pair.h:22]: @akohlmey sure, that seems ok

**Comment by sjplimp** [comment]
@rbberger yes to the shortened names.  Also these bitmasks are needed more globally than Pair.
Integrate::ev_set uses them to set eflag,vflag.  Then those vars are passed to pair, bond, angle, etc.
So all of them need the masks to decode.  So how about the masks are defined in force.h ?

**Comment by rbberger** [comment]
@sjplimp done. now what is left is updating all other code paths where those flags are used, checked and set. I'll look at this next week.

File: src/kspace.cpp
Comment (rbberger) [inline src/kspace.cpp:240]: @sjplimp if `vflag = VIRIAL_CENTROID` (=8), this will be `2` and therefore true. Not sure if that is intentional or an outdated code path.

Comment (akohlmey) [reply src/kspace.cpp:240]: This is intentional. It would have been cleaner to have written `vflag_atom = vflag & 12;`
The centroid description of the per-atom stress is equivalent to the virial description for all interactions with spherical symmetry (pair styles like lj/cut or morse). Thus all such pair styles are automatically compatible and that also includes KSpace. Only bonded interactions and many-body potentials need a different derivation of the per-atom stress (for global stress the result is the same again for both representations).

Comment (rbberger) [reply src/kspace.cpp]: @akohlmey thanks for the clarification.

**Comment by rbberger** [comment]
@sjplimp all yours.

File: src/CLASS2/pair_lj_class2.cpp
Comment (sjplimp) [inline src/CLASS2/pair_lj_class2.cpp:38]: this is the default - the point of changing the default was to remove this lines from all the true pairwise pair styles

**Comment by sjplimp** [comment]
@rbberger back to you.  This will require adding the CENTROID enum to compile. @donatas-surblys I think I added all the restrictions we talked about.  Can you take a look at the bottom of the doc/src/compute_stress_atom.rst doc page to see if it all looks good?  Ditto for src/compute_centroid_stress_atom.cpp

**Comment by rbberger** [comment]
@sjplimp the CENTROID enum was already added in `force.h`. If you want to change the values from 1, 2, 4, you can do it there. The code doesn't compile for various other reasons. Let me know when you are done pushing and change the assignment to me.

File: src/compute_centroid_stress_atom.cpp
Comment (rbberger) [inline src/compute_centroid_stress_atom.cpp:259]: @sjplimp since `kspace->cvatom` doesn't exist, it's unclear what your intentions are with this change.

**Comment by sjplimp** [comment]
@rbberger ok, builds now.  See if you agree with all the logic?

**Comment by rbberger** [comment]
@sjplimp @akohlmey Only found a minor issue. Otherwise, looks good to me. Nothing showed up in the regression testing. Since this a critical change, I'm asking for at least 2 core dev approvals before merging (myself excluded since I created the PR).

File: src/compute_centroid_stress_atom.cpp
Comment (donatas-surblys) [inline src/compute_centroid_stress_atom.cpp:263]: There will probably never be a centroid stress definition for kspace, and even if there was it would probably by a symmetric tensor, so the (currently commented out) code is a bit  misleading. Also, from a pure convenience point of view, if someone wanted to force the use of the original kspace stress definition for centroid/stress/atom (which is OK for total system heat flux), he would need to not only change the centroidstressflag, but also modify this part. My personal preference would be to revert the code or at least leave the original code that uses vatom commented out.

**Comment by donatas-surblys** [comment]
I've looked over the documentation and the code in centroid_stress_atom.cpp. Except for the one comment, everything looks OK.

**Comment by sjplimp** [comment]
@rbberger I will address @donatas-surblys comment in the next few minutes

**Comment by akohlmey** [comment]
> @rbberger I will address @donatas-surblys comment in the next few minutes

@sjplimp this fails to compile now.

**Comment by rbberger** [comment]
@sjplimp I can't merge this without approval. I need either yours or Axel to do it.

**Review by sjplimp — APPROVED** [review]

**Pull request merged**
